src/rust/ide/src/model/suggestion_database/entry.rs

```diff
@@ -0,0 +1,426 @@
+                , Entry::apply_opt_field_update("documentation", &mut self.documentation, documentation)
+                , self.apply_scope_update(scope)
+            ];
+        let other_update_results = SmallVec::from_buf(other_update_results).into_iter();
+        let other_update_errors  = other_update_results.filter_map(|res| res.err());
+        let arg_update_errors    = arguments.into_iter().flat_map(|arg| self.apply_arg_update(arg));
+        arg_update_errors.chain(other_update_errors).collect_vec()
+    }
+
+    fn apply_arg_update(&mut self, update:language_server::types::SuggestionArgumentUpdate)
+                        -> Vec<failure::Error> {
+        use language_server::types::SuggestionArgumentUpdate as Update;
+        let error = |index| {
+            let name = self.name.clone();
+            vec![failure::Error::from(InvalidArgumentIndex {name,index})]
+        };
+        match update {
+            Update::Add {index,..} if index > self.arguments.len() => {
+                error(index)
+            }
+            Update::Remove {index} | Update::Modify {index,..} if index >= self.arguments.len() => {
+                error(index)
+            }
+            Update::Add {index,argument} => {
+                self.arguments.insert(index,argument);
+                vec![]
+            }
+            Update::Remove {index} => {
+                self.arguments.remove(index);
+                vec![]
+            }
+            Update::Modify {index,name,repr_type,is_suspended,has_default,default_value} => {
+                let arg     = &mut self.arguments[index];
+                type E = Entry;
+                let results =
+                    [E::apply_field_update    ("name"         ,&mut arg.name         ,name)
+                        ,E::apply_field_update    ("repr_type"    ,&mut arg.repr_type    ,repr_type)
+                        ,E::apply_field_update    ("is_suspended" ,&mut arg.is_suspended ,is_suspended)
+                        ,E::apply_field_update    ("has_default"  ,&mut arg.has_default  ,has_default)
+                        ,E::apply_opt_field_update("default_value",&mut arg.default_value,default_value)
+                    ];
+                SmallVec::from_buf(results).into_iter().filter_map(|res| res.err()).collect_vec()
+            }
+        }
+    }
+
+    fn apply_scope_update
+    (&mut self, update:Option<FieldUpdate<language_server::types::SuggestionEntryScope>>)
+     -> FallibleResult {
+        if let Some(update) = update {
+            let err = || Err(failure::Error::from(InvalidFieldUpdate("scope")));
+            match &mut self.scope {
+                Scope::Everywhere       => { return err() },
+                Scope::InModule {range} => {
+                    if let Some(value) = update.value { *range = value.into() }
+                    else                              { return err()          }
+                }
+            }
```

Comment:
do we need to use `return` ? Id prefer not to use early exit points in our codebase

Reply:
Fixed, however it's almost like having ? which we widely accept.